hathor/nanocontracts/storage/patricia_trie.py

```diff
@@ -0,0 +1,389 @@
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import hashlib
+from dataclasses import dataclass, field
+from itertools import chain
+from typing import Iterable, NamedTuple, NewType, Optional
+
+from hathor.nanocontracts.storage.backends import NodeTrieStore
+
+NodeId = NewType('NodeId', bytes)
+
+
+class DictChildren(dict[bytes, NodeId]):
+    """Data structure to store children of tree nodes."""
+    def find_prefix(self, a: bytes) -> Optional[tuple[bytes, NodeId]]:
+        """Find the key that is a prefix of `a`."""
+        # TODO Optimize search.
+        for key, node_id in self.items():
+            if a.startswith(key):
+                return key, node_id
+        return None
+
+    def copy(self):
+        """Return a copy of itself."""
+        return DictChildren(self)
+
+
+@dataclass(kw_only=True, slots=True)
+class Node:
+    """This is a node in the Patricia trie.
+
+    Each node can carry an object or not. If a node does not carry an object, its key has never been directly added to
+    the trie but it was created because some keys have the same prefix.
+
+    Note: We might be able to remove the length.
+    """
+
+    key: bytes
+    length: int
+    content: Optional[bytes] = None
+    children: DictChildren = field(default_factory=DictChildren)
+    _id: Optional[NodeId] = None
+
+    @property
+    def id(self) -> NodeId:
+        assert self._id is not None
+        return self._id
+
+    def copy(self, content: Optional[bytes] = None, children: Optional[DictChildren] = None) -> 'Node':
+        """Generate a copy of this node except by the id field."""
+        content = content if content is not None else self.content
+        children = children if children is not None else self.children.copy()
+        return Node(key=self.key, length=self.length, content=content, children=children)
+
+    def calculate_id(self) -> NodeId:
+        """Calculate a merkle hash to serve as node id.
+
+        This method assumes that all children already have their ids calculated.
+        """
+        h = hashlib.sha1()
+        h.update(self.key)
+        if self.content is not None:
+            h.update(self.content)
+        sorted_child_ids = sorted(list(self.children.values()))
+        for child_id in sorted_child_ids:
+            h.update(child_id)
+        return NodeId(h.digest())
+
+    def update_id(self) -> None:
+        """Update node id."""
+        assert self._id is None
+        self._id = self.calculate_id()
+
+
+class IterDFSNode(NamedTuple):
+    """Item yielded by `PatriciaTrie.iter_dfs()`."""
+    node: Node
+    height: int
+    is_leaf: bool
+
+
+class PatriciaTrie:
```

Comment:
Let's create a task to review this algorithm once again before going to the mainnet.